tests/utils/tensorflow/test_callback.py

```diff
@@ -0,0 +1,71 @@
+from typing import Dict, Text, List
+from pathlib import Path
+
+import pytest
+
+from rasa.core.agent import Agent
+from rasa.core.policies.ted_policy import TEDPolicy
+from rasa.utils.tensorflow.callback import RasaModelCheckpoint
+
+
+@pytest.mark.parametrize(
+    "logs, improved",
+    [
+        (
+            [{"val_i_acc": 0.5, "val_f1": 0.5}, {"val_i_acc": 0.65, "val_f1": 0.7}],
+            True,
+        ),  # both improved
+        (
+            [{"val_i_acc": 0.54, "val_f1": 0.5}, {"val_i_acc": 0.54, "val_f1": 0.7}],
+            True,
+        ),  # one equal, one improved
+        (
+            [{"val_i_acc": 0.8, "val_f1": 0.55}, {"val_i_acc": 0.8, "val_f1": 0.55}],
+            False,
+        ),  # both equal
+        (
+            [{"val_i_acc": 0.64, "val_f1": 0.5}, {"val_i_acc": 0.41, "val_f1": 0.7}],
+            False,
+        ),  # one improved, one worse
+        (
+            [{"val_i_acc": 0.71, "val_f1": 0.35}, {"val_i_acc": 0.52, "val_f1": 0.35}],
+            False,
+        ),  # one worse, one equal
+    ],
+)
+def test_does_model_improve(logs: List[Dict[Text, float]], improved, tmpdir: Path):
+    checkpoint = RasaModelCheckpoint(tmpdir)
+    checkpoint.best_metrics_so_far = logs[0]
+    # true iff all values are equal or better and at least one is better
+    assert checkpoint._does_model_improve(logs[1]) == improved
+
+
+@pytest.fixture(scope="function")
+def trained_ted(mood_agent: Agent) -> TEDPolicy:
+    # use the moodbot agent to get a trained TEDPolicy
+    for policy in mood_agent.policy_ensemble.policies:
+        if isinstance(policy, TEDPolicy):
+            return policy
+
+
+@pytest.mark.parametrize(
+    "logs, improved",
+    [
+        ([{"val_i_acc": 0.5, "val_f1": 0.5}, {"val_i_acc": 0.5, "val_f1": 0.7}], True),
+        ([{"val_i_acc": 0.5, "val_f1": 0.5}, {"val_i_acc": 0.4, "val_f1": 0.5}], False),
+    ],
+)
+def test_on_epoch_end_saves_checkpoints_file(
+    logs: List[Dict[Text, float]], improved: bool, trained_ted: TEDPolicy, tmpdir: Path
+):
+    model_name = "checkpoint"
+    best_model_file = Path(str(tmpdir), model_name)
```

Comment:
I don't think the `str` is needed. I keep seeing it elsewhere which might mean that no-one has questioned it or that it's actually needed 😅